session/txn.go

```diff
@@ -83,8 +84,27 @@ func (txn *LazyTxn) CacheTableInfo(id int64, info *model.TableInfo) {
 func (txn *LazyTxn) init() {
 	txn.mutations = make(map[int64]*binlog.TableMutation)
 	txn.mu.Lock()
-	txn.mu.TxnInfo.State = txninfo.TxnIdle
-	txn.mu.Unlock()
+	defer txn.mu.Unlock()
+	txn.mu.TxnInfo = txninfo.TxnInfo{}
+}
+
+// call this under lock!
+func (txn *LazyTxn) updateState(state txninfo.TxnRunningState) {
+	if txn.mu.TxnInfo.State != state {
+		lastState := txn.mu.TxnInfo.State
+		lastStateChangeTime := txn.mu.TxnInfo.LastStateChangeTime
+		txn.mu.TxnInfo.State = state
+		txn.mu.TxnInfo.LastStateChangeTime = time.Now()
+		if !lastStateChangeTime.IsZero() {
+			hasLockLbl := "false"
+			if !txn.mu.TxnInfo.BlockStartTime.IsZero() {
+				hasLockLbl = "true"
+			}
+			metrics.TxnDurationHistogram.WithLabelValues(txninfo.StateLabel(lastState), hasLockLbl).Observe(time.Since(lastStateChangeTime).Seconds())
```

Comment:
I'm a bit worried about the overhead it may introduce... In past experiences, `WithLabelValues` sometimes causes noticable regressions. Using pre-defined metrics will cost less (and define `StateDurationHistogram()` and  `StateGauge()` to replace `StateLabel`...)